File: src/main/java/com/employe/Models/Employe.java
package com.employe.Models;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Transient;
import org.springframework.data.jdbc.core.mapping.AggregateReference;
import org.springframework.data.relational.core.mapping.MappedCollection;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.time.LocalDate;
import java.util.Collection;
import java.util.Collections;
import java.util.Date;
import java.util.List;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Employe implements UserDetails {
    @Id
    private Integer id;
    private String prenom;
    private String nom;
    private String motDePasse;
    private String nomUtilisateur;
    private LocalDate dateEmbauche;
    private String numeroAssuranceSocial;
    private Role role;
    private AggregateReference<Employe, Integer> superviseur;
    @MappedCollection(keyColumn = "employe")
    @Transient
    private List<HoraireQuotidien> horairesQuotidiens;
    @MappedCollection(keyColumn = "employe")
    @Transient
    private List<FeuilleDeTemps> feuillesDeTemps;
    private double tauxHoraire;

    private static String CODE_ENTREPRISE = "3052";

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return Collections.singletonList(new SimpleGrantedAuthority("ROLE_" + role.name()));
    }

    @Override
    public String getPassword() {
        return motDePasse;
    }

    @Override
    public String getUsername() {
        return nomUtilisateur;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

    public enum Role {
        ADJOINT,
        ASSOCIE,
        GERANT
    }

    public static String genererMotDePasse(String prenom, String nom) {
        return prenom.substring(0, 1).toUpperCase() + nom.substring(0, 1).toUpperCase() + CODE_ENTREPRISE;
    }

}

File: Angular/src/app/Model/Employe.ts
import {HoraireQuotidien} from "./horaire-quotidien";
import {FeuilleDeTemps} from "./feuille-de-temps";
import {Role} from "./Role";

export class Employe {

  private _id?: number;
  private _prenom: string;
  private _nom: string;
  private _nomUtilisateur: string;
  private _dateEmbauche: Date;
  private _numeroAssuranceSocial?: string;
  private _role: Role;
  private _tauxHoraire: number;
  private _superviseur?: number;

  constructor(
              prenom: string,
              nom: string,
              nomUtilisateur: string,
              dateEmbauche: Date,
              role: Role,
              tauxHoraire: number,
              id?: number,
              superviseur?: number
  ) {
    this._prenom = prenom;
    this._nom = nom;
    this._nomUtilisateur = nomUtilisateur;
    this._dateEmbauche = dateEmbauche;
    this._role = role;
    this._tauxHoraire = tauxHoraire;
    if (id) {
      this._id = id;
    }
    if (superviseur) {
      this._superviseur = superviseur;
    }
  }

  get id(): number | undefined {
    return this._id;
  }

  get prenom(): string {
    return this._prenom;
  }

  get nom(): string {
    return this._nom;
  }

  get nomUtilisateur(): string {
    return this._nomUtilisateur;
  }

  get dateEmbauche(): Date {
    return this._dateEmbauche;
  }

  get numeroAssuranceSocial(): string | undefined {
    return this._numeroAssuranceSocial;
  }

  get role(): Role {
    return this._role;
  }

  get tauxHoraire(): number {
    return this._tauxHoraire;
  }

  get superviseur(): number | undefined {
    return this._superviseur;
  }

  public static jsonToEmploye(jsonObject: any): Employe {
    let idSuperviseur = undefined;
    if (jsonObject.superviseur !== null) {
      idSuperviseur = jsonObject.superviseur.id;
    }
    const dateAvecTemp = jsonObject.dateEmbauche + "T00:00:00";
    return new Employe(
      jsonObject.prenom,
      jsonObject.nom,
      jsonObject.nomUtilisateur,
      new Date(dateAvecTemp),
      jsonObject.role,
      jsonObject.tauxHoraire,
      jsonObject.id !== null ? jsonObject.id : undefined,
      idSuperviseur
    );
  }
}
